runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/MapControlClientPool.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.control;
+
+import com.google.common.collect.Maps;
+import java.time.Duration;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * A {@link ControlClientPool} backed by a client map. It is expected that a given client id will be
+ * requested at most once.
+ */
+public class MapControlClientPool implements ControlClientPool {
+
+  /** Creates a {@link MapControlClientPool} with an unspecified timeout. */
+  public static MapControlClientPool create() {
+    return withTimeout(Duration.ofSeconds(Long.MAX_VALUE));
+  }
+
+  /**
+   * Creates a {@link MapControlClientPool} with the given timeout. Timeouts only apply to source
+   * requests.
+   */
+  public static MapControlClientPool withTimeout(Duration timeout) {
+    return new MapControlClientPool(timeout);
+  }
+
+  private final Duration timeout;
+  private final Object lock = new Object();
+
+  @GuardedBy("lock")
+  private final Map<String, CompletableFuture<InstructionRequestHandler>> clients =
```

Comment:
<!--new_thread; commit:1fd121da1417624b3b84f0300648251da64b9cb5; resolved:0-->
Why is the value type CompletableFuture, rather than simply InstructionRequestHandler? The get/put methods operate in terms of IRH, and the CompletableFuture we wrap it into seems trivial.